fix: redirect deprecated libraries resolution to the public lib DB

anmonteiro opened the pull request:
addresses this [comment](https://github.com/ocaml/dune/pull/10307#discussion_r1544964403)

anmonteiro committed:
fix: redirect deprecated libraries resolution to the public lib DB

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro commented on test/blackbox-tests/test-cases/lib-collision/lib-collision-public-same-public-name.t:
```diff
@@ -23,12 +23,12 @@ different folders.
 Without any consumers of the libraries
 
   $ dune build
```
this error becomes better: we now use the `loc` for the `(public_name ..)` field, which is the actual conflict.

jchavarri reviewed (approved):
I'm afraid I don't fully understand what the change in redirection to parent is fixing (besides the better locations in errors mentioned). But otherwise lgtm.

rgrinberg commented on src/dune_rules/lib.ml:
```diff
@@ -424,7 +424,7 @@ and resolve_result =
   | Ignore
   | Redirect_in_the_same_db of (Loc.t * Lib_name.t)
-  | Redirect of db * Lib_id.t
+  | Redirect of db * (Loc.t * Lib_name.t)
```
Isn't this constructor also used for redirecting public names to private names?

anmonteiro commented on src/dune_rules/scope.ml:
```diff
@@ -85,13 +85,13 @@ module DB = struct
     | Found lib -> Memo.return (Lib.DB.Resolve_result.found lib)
     | Deprecated_library_name lib ->
-      Memo.return (Lib.DB.Resolve_result.redirect_in_the_same_db lib)
+      Memo.return (Lib.DB.Resolve_result.redirect parent lib)
```
I believe this change is just saving us 1 redirect, because the new public name will exist in the public libs DB. We'd redirect anyway at resolution time (following 2 redirects, now just 1)

anmonteiro commented:
> Isn't this constructor also used for redirecting public names to private names?

I think this might be true. We may have to specify whether we're redirecting to a name or a library ID.

rgrinberg commented:
Just to make sure we're on the same page, I think we need two constructors:

1. `Redirect of Lib_name.t * DB.t` used for implemented `(deprecated_library_stanza ..)`. The target db here should be the public db
2. `Redirect of Lib_id.t` used for implementing public to private redirects.

anmonteiro committed:
add a new `Redirect_to_lib_id` constructor

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro commented:
OK I added a new `Redirect_to_lib_id` constructor that handles the redirection for local stanzas.

I also added `visibility` in `Lib_id` to get the "best loc" for a stanza. This way we make sure to keep showing the correct loc for the public name clash case in tests.

rgrinberg commented on src/dune_rules/lib.mli:
```diff
@@ -100,7 +100,8 @@ module DB : sig
     val found : Lib_info.external_ -> t
     val to_dyn : t Dyn.builder
-    val redirect : db -> Lib_id.t -> t
+    val redirect : db -> Loc.t * Lib_name.t -> t
```
How about renaming this one too for symmetry? E.g. `redirect_by_name`

rgrinberg commented on src/dune_rules/lib.mli:
```diff
@@ -100,7 +100,8 @@ module DB : sig
     val to_dyn : t Dyn.builder
-    val redirect : db -> Lib_id.t -> t
+    val redirect : db -> Loc.t * Lib_name.t -> t
+    val redirect_to_lib_id : db -> Lib_id.t -> t
```
`redirect_by_id` reads a bit better to me

rgrinberg commented on src/dune_rules/lib_id.ml:
```diff
@@ -18,7 +23,7 @@ module Local = struct
       | x -> x
     ;;
```
I missed this from before, but this seems wrong. How is the comparison function ignoring the other elements?

anmonteiro replied:
My thought was that, since this refers to `Local` library IDs, the `loc` would be more than enough to distinguish different lib_ids.

anmonteiro replied:
The only field we're not comparing is `enabled_if`. We could, but I couldn't come up with a `compare` function for `Blang.t` -- there's only `Blang.equal`.

rgrinberg replied:
> My thought was that, since this refers to Local library IDs, the loc would be more than enough to distinguish different lib_ids.

In that case, I would drop the `enabled_if` from the id as well.

rgrinberg commented on src/dune_rules/scope.ml:
```diff
@@ -85,13 +85,13 @@ module DB = struct
     | Found lib -> Memo.return (Lib.DB.Resolve_result.found lib)
     | Deprecated_library_name lib ->
-      Memo.return (Lib.DB.Resolve_result.redirect_in_the_same_db lib)
+      Memo.return (Lib.DB.Resolve_result.redirect parent lib)
```
Is `parent` always the public library database here?

anmonteiro replied:
yes. I'll move these functions more closely together and make that explicit in a comment

anmonteiro replied:
it adds a bit of whitespace diff noise, but I did the above in https://github.com/ocaml/dune/pull/10354/commits/3e730834ed182f5bfaed4cc16c904abd23f795c7

anmonteiro committed:
refactor: rename Redirect_to_lib_id to Redirect_by_id, Redirect_by_name

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro committed:
refactor: in scope.ml, move helper functions to the lexical scope where
they're used

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

anmonteiro committed:
revert visibility change

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

rgrinberg commented on src/dune_rules/scope.ml:
```diff
@@ -73,131 +73,133 @@ module DB = struct
+  let create_db_from_stanzas =
+    (* Here, [parent] is always the public_libs DB. Check the call to
+       [create_db_from_stanzas] below. *)
+    let resolve_found_or_redirect ~parent fr =
```
How about just rename `~parent` to `~public_db` then?

anmonteiro committed:
refactor: rename `parent` to `public_libs`

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>
